### DISCUSSION TITLE
How to disable LSP when opening a file that isn't under the cwd

### DISCUSSION THREAD

**Discussion by IceSentry** [neovim/neovim#30827] [Q&A]
When working with rust, it's pretty common for me to go to a definition of a dependency. The issue with this is that as soon as I go to the file containing the definition neovim spins up a new rust-analyzer process for that dependency. This is problematic when the dependency is massive because rust-analyzer can get very slow in those situations. What I'd like to see is the LSP only running for the current working directory, so if I open a file that isn't in the cwd it shouldn't start a new LSP process. Is this possible?

**Comment by telemachus** [comment]
You could do this with some conditional logic in an autocommand. See https://neovim.io/doc/user/autocmd.html and https://neovim.io/doc/user/api.html#nvim_create_autocmd().

**Reply by IceSentry** [reply]
The issue is more about understanding the LSP than the concept of autocommands. I have this code that runs on LspAttach. is_current_file_in_cwd() is just a simple check that makes sure the path matches and it works correctly. I just can't seem to figure out how to force the LSP to stop running for that new buffer.

```lua
	vim.api.nvim_create_autocmd('LspAttach', {
        group = vim.api.nvim_create_augroup('kickstart-lsp-attach', { clear = true }),
        callback = function(event)
          if not is_current_file_in_cwd() then
            print('Current file is not in cwd. Detaching LSP. client_id: ' .. event.data.client_id)
            vim.lsp.buf_detach_client(event.buf, event.data.client_id)
          end
		end,
	})
```

`buf_detach_client` doesn't seem to be doing anything and rust_analyzer keeps trying to index the file. I'm really surprised by this behaviour honestly, vscode doesn't do that at least. Maybe I'm just too new to neovim.

**Reply by IceSentry** [reply]
I think I figured it out, I need to use vim.defer_fn based on this comment https://github.com/neovim/nvim-lspconfig/issues/2626#issuecomment-2117022664

It actually works, but rust-analyzer itself seems to ignore it. So I might need to make it rust specific and not just LSP.

**Comment by telemachus** [comment]
> The issue is more about understanding the LSP than the concept of autocommands.

Yes, I agree. I meant that you can use an autocommand *once you've read more about neovim and how it handles LSPs*. At the moment, you probably have a plugin that automatically (and unconditionally) starts the rust server and attaches it anytime you open a file in rust. I don't know if you can get exactly what you want with whatever plugins you are using. However, you can control LSPs without a plugin. If you go that route, you can write your own autocommand and only start and attach the rust server if the file is in your current working directory. That's obviously more work, but it can be done. See links below.

+ https://neovim.io/doc/user/lsp.html#_quickstart
+ https://vonheikemen.github.io/devlog/tools/neovim-lsp-client-guide

> `buf_detach_client` doesn't seem to be doing anything and rust_analyzer keeps trying to index the file. I'm really surprised by this behaviour honestly, vscode doesn't do that at least. Maybe I'm just too new to neovim.

This may be more about rust_analyzer than `buf_detach_client`. See [the note in the docs for `buf_detach_client`](https://neovim.io/doc/user/lsp.html#vim.lsp.buf_detach_client()).

> Detaches client from the specified buffer. **Note**: While the server is notified that the text document (buffer) was closed, it is still able to send notifications should it ignore this notification.

If you don't want to handle the LSP entirely on your own, and `buf_detach_client` doesn't work, you can also kill the server with [`vim.lsp.stop_client()`](https://neovim.io/doc/user/lsp.html#vim.lsp.stop_client()). (You may need to use the `force` option.) That has its own trade-offs. In particular, you may want to keep the server running for other buffers.

**Reply by IceSentry** [reply]
I should have specified, but I'm using kickstart.nvim as a template and I pretty much didn't touch the default config of the lsp other than enabling rust_analyzer and the code I shared above that I added to the LspAttach event.

Looks like `mason-lspconfig` is probably the thing that sets up the file association and starts it automatically. I was using kickstart because I didn't want to have to think about everything and have something ready to go with sane defaults, but clearly that was a mistake.

> [vim.lsp.stop_client()](https://neovim.io/doc/user/lsp.html#vim.lsp.stop_client())

I actually did try to stop the client, but the client_id given by the LspAttach event was the same id as the client from the original buffer so stopping it would also stop the client for the original buffer. 

I guess I'll just have to configure everything myself instead of trying to use pre-configured plugins. I guess that's the whole reason I was trying to get into neovim in the first place. I wanted to have more control over my editor, but I've been faced with so many tiny missing features that I feel like I'm spending more time configuring things than actually programming. At least I got way more familiar with the bindings in the process.
